Page view approach for economic recovery subgraph

st-alice opened the pull request:
# Summary

A PR related to the page view approach for the economic recovery subgraph. 

Before running this notebook, please make sure you have created the subgraph by running: 
- `step_one_identify_seed_pages.ipynb`
- `step_two_extract_seed_sessions.sql`
- `step_three_extract_nodes_and_edges.sql`
- `step_four_create_networkx_graph.ipynb`

There are two notebooks to review:
- `page_view_approach_economic_recovery` - filters the subgraph to a final list of pages related to the economic recovery whole user journey
- `page_view_approach_economic_recovery_visualisation` - plots a bokeh graph and defines seed0 and seed1 pages to explore their neighbours

# Checklists

<!--
These are do-confirm checklists; it confirms that you have DOne each item.

Outstanding actions should be completed before reviewers are assigned; if actions are
irrelevant, please try and add a comment stating why.

Incomplete pull/merge requests may be blocked until actions are resolved, or closed at
the reviewers' discretion.
-->

This pull/merge request meets the following requirements:

- [X] code runs
- [X] [developments are ethical][data-ethics-framework] and secure
- [X] you have made proportionate checks that the code works correctly
- [ ] test suite passes
- [X] developments adhere to AQA plan (see `docs/aqa/aqa_plan.md`)
- [ ] data log updated (see `docs/aqa/data_log.md`), if necessary
- [X] assumptions, and caveats log updated (see `docs/aqa/assumptions_caveats.md`), if
  necessary
- [X] [minimum usable documentation][agilemodeling] written in the `docs` folder

Comments have been added below around the incomplete checks.

[agilemodeling]: http://agilemodeling.com/essays/documentLate.htm
[data-ethics-framework]: https://www.gov.uk/government/publications/data-ethics-framework

st-alice committed:
Added functions and imports for page view approach

st-alice committed:
Created pre-defined economic recovery subgraph

st-alice committed:
Explored page view approach for economic recovery. Includes investigation of distribution of session hit data

st-alice committed:
Updated documentation for exploration of page view appraoch

st-alice committed:
Added code to improve the precision of the subgraph

st-alice committed:
Improved precision of the subgraph

st-alice committed:
Added code to determine ranking score for final set of pages

st-alice committed:
Updated documentation

st-alice committed:
Removed duplicate file

st-alice committed:
Looked at adjacent neighbours to identify communities within a WUJ

st-alice committed:
Added Bokeh plot to visualise neighbourhood

st-alice committed:
Fixed typo

st-alice committed:
Added additional parameters to output

JakeRutherford commented on notebooks/page view functional network/page_view_approach_economic_recovery.ipynb:
```diff
@@ -0,0 +1,1508 @@
+    "import pandas as pd\n",
+    "import matplotlib.pyplot as plt\n",
+    "from tqdm.notebook import tqdm\n",
+    "import operator\n",
+    "from operator import itemgetter\n",
```
You write `import operator` and `itemgetter from operator`, but later on in your code, you write `operator.itemgetter`, when you could have just written `itemgetter` without the `operator.` part before it. See a comment later on for more detail. Besides this, you don't use operator for anything else, so you should just be able to delete 'import operator' and keep `from operator import itemgetter`

st-alice replied:
Thanks for spotting this Jake! Resolved in commit 287f84a

JakeRutherford commented on notebooks/page view functional network/page_view_approach_economic_recovery.ipynb:
```diff
@@ -0,0 +1,1508 @@
+    "# higher values indicate higher centrality\n",
+    "g_di = nx.DiGraph(g)\n",
+    "g_di_between = nx.betweenness_centrality(g_di)\n",
+    "g_between = dict(sorted(g_di_between.items(), key=operator.itemgetter(1),reverse=True))\n",
```
Since you wrote `from operator import itemgetter` earlier on in the code, you can just write `key=itemgetter(1)` here.
But, you could also write it as this:
`key = lambda x: x[1]`
itemgetter should be a bit faster than the lambda function though.

st-alice replied:
Thanks again, resolved in commit 287f84a

JakeRutherford commented on notebooks/page view functional network/page_view_approach_economic_recovery.ipynb:
```diff
@@ -0,0 +1,1508 @@
+    "# degree centrality: counts the number of incoming and outgoing relationships from a node - 'most connected'\n",
+    "# the higher the degree, the more central the node is\n",
+    "g_degree = nx.degree_centrality(g)\n",
+    "g_degree = dict(sorted(g_degree.items(), key=operator.itemgetter(1),reverse=True))\n",
```
Same as above

st-alice replied:
Similarly resolved in commit 287f84a

st-alice committed:
Removed redundant operator import

st-alice committed:
Fixed typo

JakeRutherford commented on notebooks/page view functional network/page_view_approach_economic_recovery.ipynb:
```diff
@@ -0,0 +1,1507 @@
+   "outputs": [],
+   "source": [
+    "# sum all pages shortest page paths of both `seed0` nodes to get an overall shortest page path proxy metric\n",
+    "combined_paths = []\n",
+    "\n",
+    "for page, closeness in finding_job.items():\n",
+    "    for page1, closeness1 in apprentice.items():\n",
+    "        if page == page1: \n",
+    "            combined_score = closeness + closeness1\n",
+    "            combined_paths.append(page)\n",
+    "            combined_paths.append(combined_score)\n",
+    "            \n",
+    "combined_paths=dict(zip(combined_paths[::2], combined_paths[1::2]))\n",
+    "combined_paths={k: v for k, v in sorted(combined_paths.items(), key=lambda item: item[1], reverse=True)}"
```
I wonder if there is a cleaner way to write this. It looks like you have two dicts: finding_jobs and apprentice. You're creating a new dict, in which you're finding the keys they have in common, then summing the values in each dict, where the keys match. Here is some code that does this in one line.

`set(a) & set(b)` find all the keys in common. Then it's just a case of indexing into the two dicts at the location of those keys in common and adding the values.

```
a = {'a': 3, 'd': 9, 'h': 1}
b = {'a': 1, 'h': 7, 'p': 5}

{k: a[k] + b[k] for k in set(a) & set(b)}
```
Output: `{'a': 4, 'h': 8}`

JakeRutherford commented on notebooks/page view functional network/page_view_approach_economic_recovery.ipynb:
```diff
@@ -0,0 +1,1507 @@
+   "source": [
+    "combined_paths = []\n",
+    "\n",
+    "for page, closeness in finding_job.items():\n",
+    "    for page1, closeness1 in apprentice.items():\n",
+    "        if page == page1: \n",
+    "            combined_score = closeness + closeness1\n",
+    "            combined_paths.append(page)\n",
+    "            combined_paths.append(combined_score)\n",
+    "            \n",
+    "combined_paths=dict(zip(combined_paths[::2], combined_paths[1::2]))\n",
+    "combined_paths={k: v for k, v in sorted(combined_paths.items(), key=lambda item: item[1], reverse=True)}\n",
```
I wonder if there is a cleaner way to write this. It looks like you have two dicts: finding_jobs and apprentice. You're creating a new dict, in which you're finding the keys they have in common, then summing the values in each dict, where the keys match. Here is some code that does this in one line.

`set(a) & set(b)` find all the keys in common. Then it's just a case of indexing into the two dicts at the location of those keys in common and adding the values.

```
a = {'a': 3, 'd': 9, 'h': 1}
b = {'a': 1, 'h': 7, 'p': 5}

{k: a[k] + b[k] for k in set(a) & set(b)}
```
Output: `{'a': 4, 'h': 8}`

st-alice replied:
Nice! Refactored code in commit 0ca3b73

JakeRutherford commented on notebooks/page view functional network/page_view_approach_economic_recovery.ipynb:
```diff
@@ -0,0 +1,1507 @@
+   "outputs": [],
+   "source": [
+    "# sort nodes in descending order\n",
+    "finding_job={k: v for k, v in sorted(shortest_paths_seed0_1.items(), key=lambda item: item[1], reverse=True)}"
```
Very nice dict comprehension. Not worth rewriting this, but just a tip to store in your mind:
`finding_job = {k: v for k, v in sorted(shortest_paths_seed0_1.items(), key = lambda item: -item[1])}` is equivalent to this code. Notice the `reverse=True` is removed and a minus sign comes before `item[1]`.

JakeRutherford commented on notebooks/page view functional network/page_view_approach_economic_recovery.ipynb:
```diff
@@ -0,0 +1,1507 @@
+    "        docType == 'news_story' or docType == 'press_release' or docType == 'world_location_news_article' or\n",
+    "        docType == 'world_news_story' or docType == 'fatality_notice' or docType == 'tax_tribunal_decision' or\n",
+    "        docType == 'utaac_decision' or docType == 'asylum_support_decision' or docType == 'employment_appeal_tribunal_decision' or\n",
+    "        docType == 'employment_tribunal_decision' or docType == 'service_standard_report' or docType == 'cma_case' or\n",
+    "        docType == 'decision' or docType == 'oral_statement' or docType == 'written_statement' or\n",
+    "        docType == 'authored_article' or docType == 'correspondence' or docType == 'speech' or \n",
+    "        docType == 'government_response' or docType == 'case_study'): \n",
+    "        document_type_dict[docType] = 'news and communication'\n",
+    "    \n",
+    "    elif (docType == 'completed_transaction' or docType == 'local_transaction' or docType == 'form' or \n",
+    "          docType == 'calculator' or docType == 'smart_answer' or docType == 'simple_smart_answer' or \n",
+    "          docType == 'place' or docType == 'licence' or docType == 'step_by_step_nav' or \n",
+    "          docType == 'transaction' or docType == 'answer' or docType == 'guide'):\n",
+    "          document_type_dict[docType] = 'services'\n",
+    "    \n",
+    "    elif (docType == 'regulation' or docType == 'detailed_guide' or docType == 'manual' or \n",
+    "          docType == 'manual_section' or docType == 'guidance' or docType == 'map' or \n",
+    "          docType == 'calendar' or docType == 'statutory_guidance' or docType == 'notice' or \n",
+    "          docType == 'international_treaty' or docType == 'travel_advice' or docType == 'promotional' or \n",
+    "          docType == 'international_development_fund' or docType == 'countryside_stewardship_grant' or docType == 'esi_fund' or \n",
+    "          docType == 'business_finance_support_scheme' or docType == 'statutory_instrument' or docType == 'hmrc_manual' or \n",
+    "          docType == 'standard'):\n",
+    "          document_type_dict[docType] = 'guidance_and_regulation'\n",
+    "\n",
+    "    elif (docType == 'impact_assessment' or docType == 'policy_paper' or docType == 'open_consultation' or \n",
+    "          docType == 'closed_consultation' or docType == 'consultation_outcome'):\n",
+    "          document_type_dict[docType] = 'policy_and_engagement'    \n",
+    "\n",
+    "    elif (docType == 'dfid_research_output' or docType == 'independent_report' or docType == 'research' or \n",
+    "          docType == 'statistics' or docType == 'national_statistics' or docType == 'statistics_announcement' or \n",
+    "          docType == 'national_statistics_announcement' or docType == 'official_statistics_announcement' or \n",
+    "          docType == 'statistical_data_set' or docType == 'official_statistics'):\n",
+    "          document_type_dict[docType] = 'research_and_statistics'    \n",
+    "\n",
+    "    elif (docType == 'transparency' or docType == 'corporate_report' or docType == 'foi_release' or \n",
+    "          docType == 'aaib_report' or docType == 'raib_report' or docType == 'maib_report'):\n",
+    "          document_type_dict[docType] = 'transparency'    \n",
+    "    \n",
+    "    else: \n",
+    "        document_type_dict[docType] = 'other' \n",
```
That's a lot of `or`s! Unless you really enjoy writing the word `or`, you don't need to do this. You could do something like this instead:

```
# define a set for news and communication doc types
news_and_comms_doctypes = {
    'medical_safety_alert', 'drug_safety_update', 'news_article',
    'news_story', 'press_release', etc...
}

# define a set for service doc types
service_doctypes = {
    ...
}

# and so on for the others
```

Then do this:
```
for docType, docSupertype in document_type_dict.items():
    if docType in news_and_comms_doctypes: 
        document_type_dict[docType] = 'news and communication'
    
    elif docType in service_doctypes:
        document_type_dict[docType] = 'services'

    # and so on...
```

st-alice replied:
Aha, much nicer! Thank you Jake. Resolved in commit 82c2170

JakeRutherford reviewed (commented):
Just a few things to consider tweaking. Definitely seeing improvements in your Python skills, well done :)

Maybe now is a good time to start thinking about your coding style.
E.g. look at this:

```
a = [1,2,3,4]
b=[5, 2, 9, 10]
```
and
```
a = [1, 2, 3, 4]

b = [5, 2, 9, 10]
```

Which one looks more pleasing to your eyes? Hopefully the second one. Can read more tips on writing pretty Python code that conforms to the industry standard style guide here: https://www.python.org/dev/peps/pep-0008/

st-alice committed:
Refactored code

st-alice committed:
Refatored code to remove multiple ORs

st-alice committed:
Fixed typos

st-alice commented:
Thanks @JakeMSc for your thorough and handy review as always. I have addressed the comments, and refactored the code where indicated. Also, many thanks for passing on the standard style guide - really useful resource!

JakeRutherford reviewed (approved):
"Felix and Jake approve!"
-Felix
